[tivo] Update tivo binding for OH3

mlobstein opened the pull request:
The goal of this PR is to finish the overhaul of the Tivo binding for OH2 started by @AndyXMB and bring it forward to OH3. The discussion of this binding for OH2 was here: https://community.openhab.org/t/tivo-1-1-protocol-new-binding-contribution/5572

Test build jar files here:
https://github.com/mlobstein/mlobstein-beta-test/raw/master/org.openhab.binding.tivo-3.1.0-SNAPSHOT.jar
https://github.com/mlobstein/mlobstein-beta-test/raw/master/org.openhab.binding.tivo-3.1.0-SNAPSHOT-sources.jar

Bonus jars back-ported to run under 2.5.x:
https://github.com/mlobstein/mlobstein-beta-test/raw/master/org.openhab.binding.tivo-2.5.11-SNAPSHOT.jar
https://github.com/mlobstein/mlobstein-beta-test/raw/master/org.openhab.binding.tivo-2.5.11-SNAPSHOT-sources.jar

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,491 @@
+
+        try {
+            if (streamReader != null) {
+                while (streamReader.isAlive()) {
+                    streamReader.stopReader();
+                }
```
Stopping of the Thread seems quite cumbersome. You could use `Thread.join(timeout)` here. That would also make the CountDownLatch unnecessary.

mlobstein replied:
Fixed by removing the stopReader() method an moving the logic up to connSocketDisconnect()

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,491 @@
+            logger.debug(" doNappTime '{}' - I feel like napping for '{}' milliseconds",
+                    tivoConfigData.getCfgIdentifier(), tivoConfigData.getCfgCmdWait());
+            TimeUnit.MILLISECONDS.sleep(tivoConfigData.getCfgCmdWait());
+        } catch (Exception e) {
+        }
```
Better throw the InterruptedException to finish the thread fast.

mlobstein replied:
Why throw InterruptedException during the sleep?

fwolter replied:
It is raised when OH stops. In this case you need to exit the current program flow on the shortest path.

mlobstein replied:
same comment as above

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,491 @@
+         * @throws IOException
+         */
+        public StreamReader(InputStream inputStream) {
+            this.setName("OH-binding-" + tivoConfigData.getCfgIdentifier() + "-" + tivoConfigData.getCfgHost() + "."
+                    + tivoConfigData.getCfgTcpPort());
```
The thread name should be `OH-binding-<ThingUID>-...`

mlobstein replied:
cfgIdentifier is populated in the handler like this:
tivoConfigData.setCfgIdentifier(String.valueOf(getThing().getUID()));

But I will change it to use a local field to avoid confusion.

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,491 @@
+        public void run() {
+            try {
+                logger.debug("streamReader {} is running. ", tivoConfigData.getCfgIdentifier());
+                while (!stopReader && !Thread.currentThread().isInterrupted()) {
```
You could invoke `Thread.interrupt()` if you want to stop the thread. That would make the flag unnecessary.

mlobstein replied:
I removed the flag

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,153 @@
+		<description>Monitor
+			and control your TiVo via DIRECT SOCKET commands leveraging the TIVO protocol 1.1 specification.
+			The TiVo TCP Control Protocol is an ASCII-based command protocol for remote control of a TiVo DVR
+			over a TCP local
```
Can you fix the line break?

mlobstein replied:
I was fighting with spotless over these. I finally figured out that it works best to make it one long line and then let it fix it. If you try and fix each line that is already broken up, it ends up like this :(

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,153 @@
+
+		<config-description>
+			<parameter name="address" type="text" required="true">
+				<default></default>
```
What's the purpose of this?

mlobstein replied:
leftover code, removed

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,153 @@
+		<config-description>
+			<parameter name="address" type="text" required="true">
+				<default></default>
+				<required>true</required>
```
The `required` tag is deprecated. You could use the attribute instead: `required="true"`

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,153 @@
+				<default>31339</default>
+				<required>true</required>
+				<label>TCP Port</label>
+				<description>The
+					TCP port number used to connect to the TiVo.<![CDATA[ <b> ]]>Default:
+					31339<![CDATA[ </b> ]]></description>
```
Line break. Same for below.

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,153 @@
+		<description>Displays the current channel number. When changed, tunes the DVR to the specified channel, cancelling any
+			recordings in progress if necessary i.e. all tuners are already in use / recording. The TiVo must be in Live TV mode
+			for this command to work. Type: Number (1-9999) [Decimals allowed for OTA sub-channels], DisplayFormat: %n</description>
+		<category>ChannelChange</category>
```
This category doesn't exist. See https://www.openhab.org/docs/concepts/categories.html#channel-categories

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,153 @@
+			recordings in progress if necessary i.e. all tuners are already in use / recording. The TiVo must be in Live TV mode
+			for this command to work. Type: Number (1-9999) [Decimals allowed for OTA sub-channels], DisplayFormat: %n</description>
+		<category>ChannelChange</category>
+		<state min="1" max="9999" step="0.1" pattern="%n"></state>
```
Same for below.
```suggestion
		<state min="1" max="9999" step="0.1" pattern="%d"></state>
```

fwolter reviewed (changes requested):
Thanks for migrating this from OH 1.x! Some comments may relate to legacy code. Would be awesome if you could address these, too.

mlobstein commented on bundles/org.openhab.binding.tivo/README.md:
```diff
@@ -0,0 +1,199 @@
+
+* The following rule shows how a string change to the item `TiVo_KeyboardStr` is split into individual characters and sent to the TiVo.  The method to send a keystroke multiple times is used to simulate rapid keystrokes required to achieve number based searched.  
+
+* A simple custom template widget can be used within the HABpanel user interface for tablet-based searches.  See [this discussion thread] (https://community.openhab.org/t/tivo-1-1-protocol-new-binding-contribution/5572/21?u=andymb).
```
All extra spaces removed

mlobstein commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,404 @@
+        value = conf.get(CONFIG_ADDRESS);
+        if (value != null) {
+            tivoConfigData.setCfgHost(String.valueOf(value));
+        }
```
done

mlobstein commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,404 @@
+                        break;
+                    case STANDBY:
+                        updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE,
+                                "STANDBY MODE: Send command TIVO to Remote Control Button (IRCODE) item to wakeup.");
```
fixed

mlobstein commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,404 @@
+        TivoStatusData tmpStatus = tivoConnection.get().getServiceStatus();
+        try {
+            // If desired channel has decimal (OTA channels), parse the sub channel number
+            if (command.contains(".")) {
+                String[] channelArray = command.split("\\.");
+                channel = Integer.valueOf(channelArray[0]).intValue();
+                subChannel = Integer.valueOf(channelArray[1]).intValue();
```
The user should be sending valid channel numbers like '105' or '4.1'. I will add IndexOutOfBoundsException to the catch block below to trap it along with NumberFormatException.

fwolter replied:
Catching IndexOutOfBoundsException should be avoided. Better check the length beforehand.

mlobstein replied:
I changed it to use a regex similar to the status parsing.

mlobstein commented on bundles/org.openhab.binding.tivo/README.md:
```diff
@@ -0,0 +1,199 @@
+* Special characters must also be changed to the appropriate command e.g. the comma symbol( ,) must not be sent it should be replaced by 'COMMA'.
+
+
+## Parameters
```
Fixed

mlobstein commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusData.java:
```diff
@@ -0,0 +1,216 @@
+ * {@link TivoStatusData} class stores the data from the last status query from the TiVo and any other errors / status
+ * codes.
+ *
+ * @param cmdOk boolean true = last command executed correctly, false = last command failed with error message
+ * @param channelNum int = channel number, -1 indicates no channel received. Valid channel range 1-9999.
+ * @param msg string status message from the TiVo socket
+ * @param pubToUI boolean true = this status needs to be published to the UI / Thing, false = do not publish (or it
+ *            already has been)
+ * @param connectionStatus ConnectionStatus enum UNKNOWN= test not run/default, OFFLINE = offline, STANDBY = TiVo is in
+ *            standby,
+ *            ONLINE = Online
```
fixed

mlobstein commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,404 @@
+            try {
+                TimeUnit.MILLISECONDS.sleep(tivoConfigData.getCfgCmdWait() * 2);
+            } catch (Exception e) {
+            }
```
I am not sure that I understand. I should catch InterruptedException instead of the generic Exception here.

fwolter replied:
Yes, but only if it's reasonable to catch it. It might be better to not catch it here to have more control over the thread finishing.

mlobstein replied:
It is a checked exception so it has to be caught somewhere. I think the intent of the catch was to just discard the exception. If it is not caught, this method and everything upstream would have to have a `thows` added.

fwolter replied:
You're completely right. The goal is to exit the current program flow at the shortest path, when an InterruptedException raises (aka OH is being stopped). So, a good approach is to throw the InterruptedException and catch it at the highest level you can. To achieve this, you need to declare `throws`.

mlobstein replied:
OK, I changed it to throw InterruptedException up to the overridden methods in the handler.

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,364 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tivo.handler;
```
Any reason this is not in the `internal` package? Same for the constants.

mlobstein replied:
fixed

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,364 @@
+        // standby mode
+        if (deviceStatus.getConnectionStatus() == ConnectionStatus.STANDBY && commandKeyword.contentEquals("TELEPORT")
+                && commandParameter.contentEquals("TIVO")) {
+            commandKeyword = "IRCODE " + commandParameter;
```
It's bad practice to re-assign method arguments. You could create a new local variable.

mlobstein replied:
done

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,364 @@
+        tivoConnection.get().connTivoDisconnect(false);
+    }
+
+    int convertValueToInt(Object value) {
+        if (value instanceof BigDecimal) {
+            return ((BigDecimal) value).intValue();
+        }
+        if (value instanceof String) {
+            return Integer.valueOf((String) value);
+        }
+        if (value instanceof Double) {
+            return ((Double) value).intValue();
+        }
+        return (Integer) value;
+    }
+
+    boolean convertValueToBoolean(Object value) {
+        return value instanceof Boolean ? ((Boolean) value) : Boolean.valueOf((String) value);
+    }
```
Are these still used?

mlobstein replied:
no, removed.

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,364 @@
+        tivoConfigData = getConfigAs(TivoConfigData.class);
+
+        tivoConfigData.setCfgIdentifier(getThing().getUID().getAsString());
+        logger.debug("TivoConfigData Obj: '{}'", tivoConfigData);
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,364 @@
+     * {@link startPollStatus} scheduled job to poll for changes in state.
+     */
+    private void startPollStatus() {
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                logger.debug("startPollStatus '{}' @ rate of '{}' seconds", getThing().getUID(),
+                        tivoConfigData.getPollInterval());
+                tivoConnection.ifPresent(TivoStatusProvider::statusRefresh);
+            }
+        };
```
Can you make this a lambda?

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/discovery/TiVoDiscoveryParticipant.java:
```diff
@@ -0,0 +1,127 @@
+            properties.put(CONFIG_HOST, inetAddress);
+            properties.put(CONFIG_PORT, port);
+
+            result = DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel("Tivo: " + label)
+                    .withProperty(CONFIG_NAME, label).withRepresentationProperty(CONFIG_NAME).build();
```
The representation property should be something more unique like the ip address.

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,475 @@
+     */
+    public void doNappTime() {
+        try {
+            logger.debug(" doNappTime '{}' - I feel like napping for '{}' milliseconds",
+                    tivoConfigData.getCfgIdentifier(), tivoConfigData.getCmdWaitInterval());
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.tivo/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,139 @@
+	</channel-type>
+	<channel-type id="menuTeleport">
+		<item-type>String</item-type>
+		<label>Change Special/Menu Screen (TELEPORT)</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -44,6 +44,9 @@
 @NonNullByDefault
 public class TivoStatusProvider {
+    private static final Pattern TIVO_STATUS_PATTERN = Pattern.compile("^CH_STATUS (\\d{4}) (?:(\\d{4}))?");
+    private static final Integer READ_TIMEOUT = 3000;
```
Primitive types (int) should be used where possible.

It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

mlobstein replied:
Fixed both and two other Integers

fwolter reviewed (approved):
LGTM

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/discovery/TiVoDiscoveryParticipant.java:
```diff
@@ -0,0 +1,127 @@
+ * @author Michael Lobstein - Updated for OH3
+ */
+@NonNullByDefault
+@Component(immediate = true, configurationPid = "discovery.tivo")
```
Does this need to be immediate?
```suggestion
@Component(configurationPid = "discovery.tivo")
```

mlobstein replied:
changed

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,478 @@
+         * @throws IOException
+         */
+        public StreamReader(InputStream inputStream) {
+            this.setName("OH-binding-" + thingUid + "-" + tivoConfigData.getHost() + "." + tivoConfigData.getTcpPort());
```
```suggestion
            this.setName("OH-binding-" + thingUid + "-" + tivoConfigData.getHost() + ":" + tivoConfigData.getTcpPort());
```

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,478 @@
+                        if (reader != null) {
+                            receivedData = reader.readLine();
+                        }
+                    } catch (SocketTimeoutException | SocketException e) {
+                        // Do nothing. Just allow the thread to check if it has to stop.
+                    }
```
You risk going into a busy loop here if exceptions continue to be thrown.

mlobstein replied:
fixed

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,478 @@
+                    String receivedData = null;
+                    try {
+                        BufferedReader reader = bufferedReader;
+                        if (reader != null) {
+                            receivedData = reader.readLine();
+                        }
```
You will busy loop if reader is null. Please avoid busy loops at all costs.

mlobstein replied:
Changed to throw exception if reader is null.

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,478 @@
+     * @throws InterruptedException
+     */
+    public void connTivoDisconnect(boolean forceDisconnect) throws InterruptedException {
+        if (forceDisconnect) {
+            connSocketDisconnect();
+        } else {
+            if (!tivoConfigData.isKeepConnActive()) {
+                doNappTime();
+                connSocketDisconnect();
+            }
+        }
```
You are performing two different behaviors based on the input boolean parameter. It would be better if you split this into two different methods instead.

mlobstein replied:
Agreed. I simplified the logic.

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,478 @@
+                }
+                this.streamReader = null;
+            }
+            if (streamWriter != null) {
+                streamWriter.close();
+                this.streamWriter = null;
+            }
+            if (tivoSocket != null) {
+                tivoSocket.close();
+                this.tivoSocket = null;
+            }
```
If closing of the writer fails, you should still attempt to close the socket.

mlobstein replied:
I moved this outside of the try block. PrintStream.close() does not throw any exceptions.

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,478 @@
+                streamReader.interrupt();
+                try {
+                    streamReader.join(TIMEOUT_SEC);
+                } catch (InterruptedException e) {
+                    logger.debug("Error joining streamReader: {}", e.getMessage());
+                }
```
An interrupt isn't an error, it is another process requesting this thread to stop. You shouldn't silently consume an InterruptException either, at the very least you should have some way to notify the caller that an interrupt occured. Usually the best way to to rethrow the exception. But in this case you should wait to rethrow it until you have attempted to close the other other stuff first. With that in mind the best option would probably be to perform your `streamReader.join` at the end and then just add the InterruptException to the method signature.

mlobstein replied:
Done

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/handler/TiVoHandler.java:
```diff
@@ -0,0 +1,372 @@
+                        tmpStatus.setChannelNum(channel);
+                        tmpStatus.setSubChannelNum(subChannel);
+                        return tmpStatus;
+                    case "CH_FAILED RECORDING":
+                        return tmpStatus;
+                    case "CH_FAILED MISSING_CHANNEL":
+                        return tmpStatus;
+                    case "CH_FAILED MALFORMED_CHANNEL":
+                        return tmpStatus;
+                    case "CH_FAILED INVALID_CHANNEL":
+                        return tmpStatus;
```
You should take advantage of switch statement fall-throughs when possible.
```suggestion
                    case "CH_FAILED RECORDING":
                    case "CH_FAILED MISSING_CHANNEL":
                    case "CH_FAILED MALFORMED_CHANNEL":
                    case "CH_FAILED INVALID_CHANNEL":
                        return tmpStatus;
```

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,459 @@
+                case "CH_FAILED NO_LIVE":
+                    return new TivoStatusData(false, -1, -1, false, "CH_FAILED NO_LIVE", true,
+                            ConnectionStatus.STANDBY);
+                case "CH_FAILED RECORDING":
+                    return new TivoStatusData(false, -1, -1, false, "CH_FAILED RECORDING", true,
+                            ConnectionStatus.ONLINE);
+                case "CH_FAILED MISSING_CHANNEL":
+                    return new TivoStatusData(false, -1, -1, false, "CH_FAILED MISSING_CHANNEL", true,
+                            ConnectionStatus.ONLINE);
+                case "CH_FAILED MALFORMED_CHANNEL":
+                    return new TivoStatusData(false, -1, -1, false, "CH_FAILED MALFORMED_CHANNEL", true,
+                            ConnectionStatus.ONLINE);
+                case "CH_FAILED INVALID_CHANNEL":
+                    return new TivoStatusData(false, -1, -1, false, "CH_FAILED INVALID_CHANNEL", true,
+                            ConnectionStatus.ONLINE);
```
```suggestion
                case "CH_FAILED RECORDING":
                case "CH_FAILED MISSING_CHANNEL":
                case "CH_FAILED MALFORMED_CHANNEL":
                case "CH_FAILED INVALID_CHANNEL":
                    return new TivoStatusData(false, -1, -1, false, rawStatus, true, ConnectionStatus.ONLINE);
```

cpmeister commented on bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java:
```diff
@@ -0,0 +1,459 @@
+     *
+     */
+    public class StreamReader extends Thread {
+        private @Nullable BufferedReader bufferedReader = null;
```
Is this thread supposed to be responsible for closing this reader?

mlobstein replied:
Yes it should. I added a method to close the reader.

AndyXMB commented:
Hi, sorry for not being able to respond earlier.  My Dad has been ill with caught Covid-19 so I had to rush away from home and have been caring for him since he came out of hospital for over 11 weeks.  Because I left in such a rush, I had no access to most of my IT systems. So thanks for picking up the work that I did on the TiVo binding and sorry for being "radio silent".  

I've just upgraded my installation today to version 3.1x and given the binding a go.  For me it fails to connect.  Put the binding into debug mode and doing some testing I've proven that the binding is connecting to the Tivo (blocks connection from a separate Telnet connection) but is not not detecting the successful connection.

Here is the relevant log segment:

`2021-01-21 15:08:22.943 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - TiVoHandler 'tivo:sckt:Living_Room' - creating

2021-01-21 15:08:22.969 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - Initializing a TiVo 'tivo:sckt:Living_Room' with config options

2021-01-21 15:08:22.973 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - Initializing a TiVo handler for thing 'tivo:sckt:Living_Room' - finished!

2021-01-21 15:08:22.975 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - Status collection 'tivo:sckt:Living_Room' will start in '5' seconds.

==> /var/log/openhab/events.log <==

2021-01-21 15:08:22.953 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING

2021-01-21 15:08:22.979 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from INITIALIZING to UNKNOWN

==> /var/log/openhab/openhab.log <==

2021-01-21 15:08:24.987 [ERROR] [io.openhabcloud.internal.CloudClient] - Error connecting to the openHAB Cloud instance

2021-01-21 15:08:25.301 [ERROR] [io.openhabcloud.internal.CloudClient] - Error connecting to the openHAB Cloud instance

2021-01-21 15:08:25.305 [INFO ] [io.openhabcloud.internal.CloudClient] - Disconnected from the openHAB Cloud service (UUID = 6aee1213-c3cc-4420-8129-b12026b55201, base URL = http://localhost:8080)

2021-01-21 15:08:27.975 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - startPollStatus 'tivo:sckt:Living_Room' @ rate of '10' seconds

2021-01-21 15:08:27.977 [DEBUG] [.internal.service.TivoStatusProvider] -  statusRefresh 'tivo:sckt:Living_Room' - EXISTING status data - 'TivoStatusData [cmdOk=false, time=Thu Jan 21 15:08:22 GMT 2021, channelNum=-1, subChannelNum=-1, msg=INITIALISING, pubToUI=false, connectionStatus=UNKNOWN]'

2021-01-21 15:08:27.979 [DEBUG] [.internal.service.TivoStatusProvider] -  connTivoConnect 'tivo:sckt:Living_Room' - starting connection process '1' of '5'.

2021-01-21 15:08:27.982 [DEBUG] [.internal.service.TivoStatusProvider] -  connSocketConnect 'tivo:sckt:Living_Room' - attempting connection to host '192.168.0.39', port '31339'

2021-01-21 15:08:27.984 [DEBUG] [.internal.service.TivoStatusProvider] -  connIsConnected 'tivo:sckt:Living_Room' - FALSE: tivoSocket=null

2021-01-21 15:08:27.985 [DEBUG] [.internal.service.TivoStatusProvider] -  connTivoSocket 'tivo:sckt:Living_Room' - requested to disconnect/cleanup connection objects

2021-01-21 15:08:27.994 [DEBUG] [.internal.service.TivoStatusProvider] -  connTivoConnect 'tivo:sckt:Living_Room' - Socket created / connection made.

2021-01-21 15:08:27.994 [DEBUG] [.internal.service.TivoStatusProvider] - streamReader tivo:sckt:Living_Room is running. 

2021-01-21 15:08:28.017 [DEBUG] [.internal.service.TivoStatusProvider] - TiVo tivo:sckt:Living_Room data received: CH_STATUS 0603 LOCAL

2021-01-21 15:08:28.018 [DEBUG] [.internal.service.TivoStatusProvider] -  statusParse 'tivo:sckt:Living_Room' - running on string 'CH_STATUS 0603 LOCAL'

2021-01-21 15:08:28.020 [DEBUG] [.internal.service.TivoStatusProvider] -  statusParse 'tivo:sckt:Living_Room' - groups 'CH_STATUS 0603 ' with group count of '2'

2021-01-21 15:08:28.021 [DEBUG] [.internal.service.TivoStatusProvider] -  statusParse 'tivo:sckt:Living_Room' - parsed channel '603'

==> /var/log/openhab/events.log <==

2021-01-21 15:08:27.989 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from UNKNOWN to OFFLINE (COMMUNICATION_ERROR): Power on device or check network configuration/connection.`

So you can see that a connection is made - channel 603 is correctly discovered, but the function that determines this (connIsConnected ?) does not appear to be working as intended / thing status not updating as a result.

I can certainly help with any testing now (back home) and any questions that anyone may have on the intent of the original work. It's an old and somewhat feature limited protocol.  But as far as I can see is still the only thing we have for Tivo devices.

If you want me to log as an issue let me know.

Andy

mlobstein commented:
Hi @AndyXMB, I think I found this issue. There is a missing null check in TivoStatusProvider on line 184. It works for me since my setup is using sub-channels. But since yours is not, the parseInt on line 185 should be skipped and is likely what is causing the exception since group 2 of the regex is null. I changed the code and updated the test build jars linked above. Can you try one of them out and see if this fixes the issue? If so, I will create a new PR.

AndyXMB commented:
Hi @mlobstein , 

Good news and a new issue...

Your fix has cured the problem when watching normal TV channels and the thing comes ONLINE as expected.  I have not tested anything else, yet as I need to set up some controls in the new UI.  I'll adapt Yanik's keypad control widget over the weekend to get something I can easily test with.

However when initially deploying your new code, I happened to be using the YouTube application on the TiVo and this does not work.  This confused me for a bit, so I have been trying to work out what is going on this evening...

When using the YouTube or AmazonPrime applications, you get the following:

```
2021-01-22 22:26:04.433 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - TiVoHandler 'tivo:sckt:Living_Room' - creating

2021-01-22 22:26:04.480 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - Initializing a TiVo 'tivo:sckt:Living_Room' with config options

2021-01-22 22:26:04.485 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - Initializing a TiVo handler for thing 'tivo:sckt:Living_Room' - finished!

2021-01-22 22:26:04.488 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - Status collection 'tivo:sckt:Living_Room' will start in '5' seconds.

==> /var/log/openhab/events.log <==

2021-01-22 22:26:04.461 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING

2021-01-22 22:26:04.486 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from INITIALIZING to UNKNOWN

==> /var/log/openhab/openhab.log <==

2021-01-22 22:26:09.488 [DEBUG] [ng.tivo.internal.handler.TiVoHandler] - startPollStatus 'tivo:sckt:Living_Room' @ rate of '10' seconds

2021-01-22 22:26:09.490 [DEBUG] [.internal.service.TivoStatusProvider] -  statusRefresh 'tivo:sckt:Living_Room' - EXISTING status data - 'TivoStatusData [cmdOk=false, time=Fri Jan 22 22:26:04 GMT 2021, channelNum=-1, subChannelNum=-1, msg=INITIALISING, pubToUI=false, connectionStatus=UNKNOWN]'

2021-01-22 22:26:09.492 [DEBUG] [.internal.service.TivoStatusProvider] -  connTivoConnect 'tivo:sckt:Living_Room' - starting connection process '1' of '5'.

2021-01-22 22:26:09.493 [DEBUG] [.internal.service.TivoStatusProvider] -  connSocketConnect 'tivo:sckt:Living_Room' - attempting connection to host '192.168.0.39', port '31339'

2021-01-22 22:26:09.495 [DEBUG] [.internal.service.TivoStatusProvider] -  connIsConnected 'tivo:sckt:Living_Room' - FALSE: tivoSocket=null

2021-01-22 22:26:09.496 [DEBUG] [.internal.service.TivoStatusProvider] -  connTivoSocket 'tivo:sckt:Living_Room' - requested to disconnect/cleanup connection objects

2021-01-22 22:26:09.503 [DEBUG] [.internal.service.TivoStatusProvider] -  connTivoConnect 'tivo:sckt:Living_Room' - Socket created / connection made.

2021-01-22 22:26:09.504 [DEBUG] [.internal.service.TivoStatusProvider] - streamReader tivo:sckt:Living_Room is running. 

==> /var/log/openhab/events.log <==

2021-01-22 22:26:09.499 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from UNKNOWN to OFFLINE (COMMUNICATION_ERROR): Power on device or check network configuration/connection.
```

When exiting the application, I get:
```
2021-01-22 22:28:12.778 [DEBUG] [.internal.service.TivoStatusProvider] - TiVo tivo:sckt:Living_Room data received: CH_STATUS 0205 LOCAL

2021-01-22 22:28:12.779 [DEBUG] [.internal.service.TivoStatusProvider] -  statusParse 'tivo:sckt:Living_Room' - running on string 'CH_STATUS 0205 LOCAL'

2021-01-22 22:28:12.782 [DEBUG] [.internal.service.TivoStatusProvider] -  statusParse 'tivo:sckt:Living_Room' - groups 'CH_STATUS 0205 ' with group count of '2'

2021-01-22 22:28:12.783 [DEBUG] [.internal.service.TivoStatusProvider] -  statusParse 'tivo:sckt:Living_Room' - parsed channel '205'

==> /var/log/openhab/events.log <==

2021-01-22 22:28:12.796 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tivo:sckt:Living_Room' changed from OFFLINE (COMMUNICATION_ERROR): Power on device or check network configuration/connection. to ONLINE

2021-01-22 22:28:12.801 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'TivoLivingRoom_TiVoStatus' changed from CH_STATUS 0147 to CH_STATUS 0205

2021-01-22 22:28:12.803 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'TivoLivingRoom_CurrentChannelForced' changed from 147 to 205

2021-01-22 22:28:12.806 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'TivoLivingRoom_CurrentChannelRequest' changed from 147 to 205
```

So immediately after you exit the application, the channel is returned and the thing goes online.  From the testing so far, if you start the binding up watching TV then you are ok, when you start the app.  But I'll need to play around a bit more to verify 100%.

This is a new issue, we had no apps when I did my original work.  Now we have YouTube, Netfix and Amazon Prime.  

I was always unhappy with the original code to set the connection status.   From your description of the fix, I guess the code is still using the channel data, returned when you first connect, to determine the connection was successful.  In the scenario when an app is being used, nothing gets returned (as verified with a manual Telnet connection).  The thing stays offline until you exit (even though the connection is established and working)

I can't remember whether the protocol even mentions apps... I'll have to find a copy and check.

It might be better to decouple the online/offline status reporting from the returning of messages.  The problem is that the socket connection is maintained when the device is in standby mode.  I originally wanted to indicate powered down/standby state, hence the original logic. In standby no messages are returned only the use of the HOME/TIVO button to wakeup etc.

There is another trick/workaround, which may help here.  If you send a deliberate garbage/invalid command, this always returns an error message response.  I had previously considered having a process that periodically sent a garbage command and detect the error message i.e. poll the connection to see if the device was still online.  The only problem is that you would want to hide these error responses from the end user etc. I could never work out how to link the sending of a message with the listener.

While I'm not convinced the above is is a good idea (hate wasting resources), we may be able to do something like:

1. Make socket connection [Success = Online or Standby] | [Fail = Powered off or invalid IP]
2. Read channel response [Success = Online] | [Fail = after timeout]
3. If Fail - send garbage command, if error message returned [Success = Online]

I'll have a play around and do some more testing over the weekend. By the way this was what the undocumented command channel was there to support ;-)

If all else fails, we document this scenario.  However I see this as the source of many 'support calls', so it would be cool if we could come up with something better.

Can I make one more suggestion.  The channel labels have lost their original reference to the protocol command i.e. **Current Channel - Forced (FORCECH)** became **Current Channel - Forced**.  I know you moved these to the description, the trouble is that the description does not display when adding items to the UI or referencing in scripts. Again this was motivated by trying to reduce support requests by making it "in your face" which commands should be sent to which channel. 

Thanks for your quick response.  Sorry for the curve ball, Andy

mlobstein commented:
Thanks for the feedback. I did not think to test with one of the streaming apps running. It is annoying that the device is basically silent when the app is running which is exactly in line with your observations. The protocol document (v1.1 - 2011) does not mention anything about interacting with apps at all. The only undocumented thing that I found was the ability to use the TELEPORT NETFLIX command to go to Netflix directly.

I think the bogus command to test the connection might be a good path to go down. I don't like the notion of setting the status to ONLINE after only just establishing a connection. I prefer to get some secondary indication that the device is responding properly before indicating that it is ONLINE.

As for the Force Channel label, it was something that got flagged during code review. It was mentioned that the rule for labels is "Labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 characters" . So that is why I shortened it to 3 words. It seemed like FORCECH was a bit of arcane info specific to the protocol that the user did not need to know, so that was the word that was dropped. If you have some suggestions for the labels to make more sense I will be happy to change them. I will add those change along with fixing the bug discussed above and adding a note to the README about the behavior you noted when running an app to a PR to get in for the 3.1.0 milestone. The binding did not make it into the 3.0.0 release, so only the snapshot build is affected right now.

AndyXMB commented:
Hi @mlobstein,

You may want to take a look at:
https://community.openhab.org/t/bogob-big-ol-grid-o-buttons-is-this-even-possible-yes-yes-it-is/115343/3?u=andymb

I tested everything in about 30 min with the BoGoB  with the exception of the volume up, volume down and mute functions.  I have these mapped to by physical TV, so I need to do some work to reset the TiVo to test these.  Everything worked with the exception of waking the TiVo up out of standby (using the TIVO / Home (as it's labelled on Virgin Media boxes) button.

I think this is caused by the same issue as above.  When the system goes into standby the state changed to OFFLINE.  The socket connection is however still in place (verified with Telnet connection refused).  I suspect your logic then does not send the TIVO command to wake the unit back up (logical it thinks it is OFFLINE).  

So not being able to wake the device up would be a real source of frustration.  Suspending and waking the TiVo when I turn on/off the TV is one of the main reasons I use the binding.  So the more I think about it the more we need a better way of trying to determine the ONLINE / OFFLINE state.  I think I had 3 states in the original code OFFLINE - socket broke, ONLINE - you need to wake me up (socket OK but no responses being returned) and ONLINE expected responses received.

Incidentally I found that the undocumented SEARCH command works to take you directly to the SEARCH page.  Our latest remotes have this button (earlier versions did not).

I may have a couple of hours tomorrow to do some more.

mlobstein commented:
Hi @AndyXMB, I did PR #9964 to fix the bugs with channel parsing, the thing being offline and the standby wake not working. Can you try these out in a recent snapshot of OH 3.0.1 to verify that the bugs are fixed?

AndyXMB commented:
Will do.  Testing in openHAB 3.1.0, Build #2175.

From: mlobstein <notifications@github.com>
Sent: 30 January 2021 22:02
To: openhab/openhab-addons <openhab-addons@noreply.github.com>
Cc: AndyXMB <andrewm.black@outlook.com>; Mention <mention@noreply.github.com>
Subject: Re: [openhab/openhab-addons] [tivo] Update tivo binding for OH3 (#9302)


Hi @AndyXMB<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAndyXMB&data=04%7C01%7C%7C0d4711bc36ba4cb639b608d8c56ab068%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637476409302565514%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=kHpvbFypcqGCZIa7VOrLO9ENS6s%2BIKv3M%2BEzxyMQ0pU%3D&reserved=0>, I did PR #9964<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenhab%2Fopenhab-addons%2Fpull%2F9964&data=04%7C01%7C%7C0d4711bc36ba4cb639b608d8c56ab068%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637476409302575509%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=0LLA2N%2FSBz3cVTj7Mt1rMhXPMt8Kzc5PKsNweG3vZXU%3D&reserved=0> to fix the bugs with channel parsing, the thing being offline and the standby wake not working. Can you try these out in a recent snapshot of OH 3.0.1 to verify that the bugs are fixed?

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenhab%2Fopenhab-addons%2Fpull%2F9302%23issuecomment-770287356&data=04%7C01%7C%7C0d4711bc36ba4cb639b608d8c56ab068%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637476409302575509%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=0g8zRePJAEG%2Bd17ir8x7QrgzJBSm%2Bvu1NoEIkLW0Mhk%3D&reserved=0>, or unsubscribe<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAF57BQRYWERRS2LCAFFLGEDS4R6WBANCNFSM4USXJUTQ&data=04%7C01%7C%7C0d4711bc36ba4cb639b608d8c56ab068%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637476409302585505%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=JELf0Lm0zZc2Cf6HxJQWGQ3Du6Uh3CZHDdbSmwNP%2BgM%3D&reserved=0>.

AndyXMB commented:
@mlobstein , tested successfully:
- standby / power on using STANDBY
- standby / power on using TIVO
- restarted binding while using an application
- channel parsing is good
- re-tested all the functions in my BoGoB

No issues found.  I'll carry on using the BoGoB to control the TiVo and keep an eye on the logs, but it's looking good.

I did try disconnecting the network cable to the TiVo. This did not update the thing status, issued commands (that failed) with again no status change.   Expected the socket connection drop to be detected and a genuine OFFLINE status, but I'm not sure what the intended behaviour is in the code now.  Appreciate that there are compromises that need to be made given the limitations of the functionality we are working with.  If you want me to re-test this, with debug logging on let me know.

I found another undocumented command FIND_REMOTE somewhere in the TiVo forums, which does what it says on the tin.

It's annoying that there does not seem to be any way to start the other apps on the TiVo.  NETFIX as you say seems to be the only app that can be started this way.  I was wondering how the "casting" functionality from my phone to the apps work e.g. YouTube.  Found that it uses the [DIscovery And Launch (DIAL) protocol](http://www.dial-multiscreen.org/).  Might be worth a 'look' for the next version ;-)

Andy

mlobstein commented:
@AndyXMB 
Thanks for verifying the fixes. I agree that the online detection could be better. From what I can see the current design is complicated by the requirement to close the connection when 'keepConActive' is set to false. Since the Tivo only allows one connection at a time, this configuration setting allows a second control application the opportunity to connect to the socket. In that scenario, the binding reporting online while the connection is closed would seem to be the correct decision.

I did find the FIND_REMOTE command as well and documented it in the sitemap example. I am thinking that there should be some better documentation for this since many users might not use sitemaps anymore. Any ideas?
